### PR TITLE
#15524 Parser

### PR DESCRIPTION
**Pull request by E1izabeth**
- [ ] grammar model
   - [x] object model
   - [ ] attributes for metadata
   - [ ] grammars read-write
 - [x] grammar graph
   - [x] object model
   - [x] algorithm of building by grammar model
 - [x] nfa by terms
   - [x] object model
   - [x] algorithm of building by grammar graph
 - [ ] analyzer-interpreter
   - [x] parsing algorithm
   - [ ] optimizations
 - [ ] applied functionality
   - [ ] finding block by parse tree
   - [ ] common SQL grammar
   - [ ] MySQL grammar (partially?)
   - [ ] ...

### CODE REVIEW

**Comment by flambchap** [comment]
Wow

**Comment by akilovich** [comment]
Usage of `java.utils.Stack` in a non-concurrent context is probably not the best idea. I suggest using `ArrayDeque` implementation of `Deque`.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java]: I think this usage is accidental.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        }
+    }
+
+    /**
+     * Find all paths from a given transition to other terminal or text-end transitions
+     * @param transition of grammar graph
+     * @return list of path ending steps
+     */
+    private List<Step> findPaths(GrammarNfaTransition transition) {
+        Stack<Step> stack = new Stack<>();
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java]: I still don't think that usage of `java.util.Stack` is a good idea. As you probably know, all class methods are synchronized and require mutex acquisition. It is inherently costly and makes this code unnecessarily slower than it could be. I suggest using an implementation of `java.util.Deque`. If this code should work in a concurrent context, then `java.util.concurrent.ConcurrentLinkedDeque` is a good choice. Otherwise, `java.util.ArrayDeque` should do.

File: product/debug/DBeaver.product.launch [hunk 3]
```diff
@@ -27,7 +27,6 @@
     <setAttribute key="selected_target_bundles">
         <setEntry value="com.github.jsqlparser@default:default"/>
         <setEntry value="com.google.gson@default:default"/>
-        <setEntry value="com.google.guava*21.0.0.v20170206-1425@default:default"/>
```
Comment (flambchap) [inline product/debug/DBeaver.product.launch:30]: Although these changes are needed in general, they do not belong to this PR.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfaOperation.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+    public Integer getExprPosition() {
+        return exprPosition;
+    }
+
+    /**
+     * Checks correctness of operation itself
+     */
+    private void validate() {
+        switch (this.kind) {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfaOperation.java]: Please indent this switch according to our conventions.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfaOperation.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            break;
+        case TERM:
+            if (this.pattern == null)
+                throw new IllegalArgumentException();
+            break;
+        case NONE:
+            // do nothing
+            break;
+        default:
+            throw new UnsupportedOperationException("Unexpected parse opeation kind " + this.kind);
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfaOperation.java]:
```suggestion
            throw new UnsupportedOperationException("Unexpected parse operation kind " + this.kind);
```

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfaOperation.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        }
+    }
+
+    @Override
+    public String toString() {
+        String result = this.kind.toString() + "#" + this.exprId + "[";
+        switch (this.kind) {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfaOperation.java]: Please format this switch according to our coding conventions.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfaBuilder.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    private int nextExprId() {
+        return this.exprId++;
+    }
+
+    /**
+     * Walk through the grammar and build the complete grammar graph in form of non-deterministic finite automaton
+     * @return start and final states of the complete grammar graph
+     */
+    public NfaFragment traverseGrammar() {
+        HashMap<GrammarRule, GrammarNfaBuilder.NfaFragment> ruleFragments = new HashMap<>();
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfaBuilder.java]: Here we can preallocate space to avoid unnecessarily reallocations of the underlying array.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfa.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ */
+package org.jkiss.dbeaver.parser.grammar.nfa;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.jkiss.dbeaver.parser.grammar.GrammarRule;
+
+/**
+ * Non-determenistic finite automaton representing complete graph of the grammar
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/nfa/GrammarNfa.java]:
```suggestion
 * Non-deterministic finite automaton representing a complete graph of the grammar
```

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/RegexExpression.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.util.regex.Pattern;
+
+/**
+ * Represents terminal part of the text with regular pattern
+ */
+public class RegexExpression extends TerminalExpression {
+    
+    // see https://docs.oracle.com/javase/tutorial/essential/regex/literals.html
+    private static final Pattern _regexEscapeCharsPattern = Pattern.compile("([\\\\\\.\\?\\!\\[\\]\\{\\}\\(\\)\\<\\>\\*\\+\\-\\=\\^\\$\\|])");
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/RegexExpression.java]:
```suggestion
    private static final Pattern _regexEscapeCharsPattern = Pattern.compile("([\\.?!\[\]{}()<>*+\-=^$|])");
```
Most of the backslashes here do not require escaping.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/ExpressionPrinter.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @Override
+    public StringBuilder visitNumber(NumberExpression numberExpression, StringBuilder sb) {
+        return this.visitUnary(numberExpression, sb);
+    }
+
+    @Override
+    public StringBuilder visitRegex(RegexExpression regexExpression, StringBuilder sb) {
+        return sb.append("\'").append(regexExpression.pattern).append("\'");
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/ExpressionPrinter.java]:
```suggestion
        return sb.append("'").append(regexExpression.pattern).append("'");
```

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParserFsmNode.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        for (ParserFsmStep s : this.steps) {
+            if (s.getPattern() != null) {
+                stepsByTerm.computeIfAbsent(s.getPattern(), p -> new ArrayList<>()).add(s);
+            } else {
+                finalSteps.add(s);
+            }
+        }
+
+        List<String> parts = new ArrayList<>(stepsByTerm.size());
+        for (Entry<String, List<ParserFsmStep>> step : stepsByTerm.entrySet()) {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParserFsmNode.java]:
```suggestion
        for (var step : stepsByTerm.entrySet()) {
```

I believe it makes it a bit easier to read. It's a matter of preference, though.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParserFsm.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.List;
+
+/**
+ * Parser finite state machine
+ */
+class ParserFsm {
+    private final List<ParserFsmNode> initialStates;
+    private final List<ParserFsmNode> allStates;
+
+    public ParserFsm(List<ParserFsmNode> initialStates, List<ParserFsmNode> parseFsmStates) {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParserFsm.java]:
```suggestion
    public ParserFsm(List<ParserFsmNode> initialStates, List<? extends ParserFsmNode> parseFsmStates) {
```

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/Parser.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    /**
+     * Parse text
+     * @param text
+     * @return result of parsing represented with discovered valid sequences of terminals
+     */
+    public ParseResult parse(String text) {
+        Deque<ParserState> queue = new ArrayDeque<>();
+        for (ParserFsmNode initialState : fsm.getInitialStates()) {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/Parser.java]: If `fsm.getInitialStates()` returns `Collection` (which is easy to do), we might preallocate memory for the Deque

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/Parser.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+     * @param ops operations to evaluate in the given context
+     * @return an updated parsing context stack state
+     */
+    private ParserStack evaluateOperations(ParserStack stack, List<GrammarNfaOperation> ops) {
+        ParserStack newStack = stack;
+
+        for (GrammarNfaOperation op : ops) {
+            // System.out.println("\t\t" + op);
+            switch (op.getKind()) {
+            case RULE_START:
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/Parser.java]: Please format this switch according to our conventions.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/Parser.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        return new ParseResult(text, grammar, results);
+    }
+
+    /**
+     * Evaluate grammar graph operations
+     * @param stack parsing context
+     * @param ops operations to evaluate in the given context
+     * @return an updated parsing context stack state
+     */
+    private ParserStack evaluateOperations(ParserStack stack, List<GrammarNfaOperation> ops) {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/Parser.java]:
```suggestion
     */
    @Nullable
    private ыефешс ParserStack evaluateOperations(ParserStack stack, List<? extends GrammarNfaOperation> ops) {
```

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseTreeNode.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+import org.jkiss.dbeaver.parser.grammar.GrammarRule;
+
+/**
+ * Parsing tree
+ */
+public class ParseTreeNode {
+    private final GrammarRule rule;
+    private final int position;
+    private final ParseTreeNode parent;
+    private final List<ParseTreeNode> childs;
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseTreeNode.java]:
```suggestion
    private final List<ParseTreeNode> children;
```

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseTreeNode.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+        return position;
+    }
+
+    public ParseTreeNode getParent() {
+        return parent;
+    }
+
+    public List<ParseTreeNode> getChilds() {
+        return childs;
+    }
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseTreeNode.java]:
```suggestion
    public List<ParseTreeNode> getChildren() {
        return children;
    }
```

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseResult.java [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+/**
+ * Parse result represented with discovered valid sequences of terminals
+ */
+public class ParseResult {
+    private final String text;
+    private final GrammarInfo grammar;
+    private final List<ParserState> results;
+    private final List<ParseTreeNode> trees;
+
+    public ParseResult(String text, GrammarInfo grammar, List<ParserState> results) {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseResult.java]:
```suggestion
    public ParseResult(String text, GrammarInfo grammar, List<? extends ParserState> results) {
```

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseResult.java [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+        ParseTreeNode current = treeRoot;
+        int pos = 0;
+        int skipDepth = 0;
+        for (ParserState state : states) {
+//            if(withWhitespaces && skipRule.getId() == state.getStep().getFrom().getId()) {
+//                continue;
+//            }
+            if (state.getStep() != null && state.getStep().getOperations() != null) {
+                for (GrammarNfaOperation op : state.getStep().getOperations()) {
+                    switch (op.getKind()) {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseResult.java]: Please format this switch according to our coding conventions.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+            return "Path of " + String.join(", ", steps);
+        }
+
+        @Override
+        public Iterator<GrammarNfaTransition> iterator() {
+            return new Iterator<GrammarNfaTransition>() {
+                private Step current = Step.this;
+
+                @Override
+                public GrammarNfaTransition next() {
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java]: This implementation violates the general contract of this method (see javadoc for `Iterator.next()`).

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+        this.terminalTransitions = terminalTransitions;
+        this.root = root;
+    }
+
+    /**
+     * Build a parser finite state machine associating grammar graph operations with terminal-guided transitions
+     * @return parser finite state machine
+     */
+    public ParserFsm buildTerminalsGraph() {
+        Map<GrammarNfaTransition, ParserFsmNode> states = new HashMap<>();
```
Comment (flambchap) [inline plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java]: We can preallocate the space in the map since we know how much memory we need.

**Review by flambchap — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/RegexExpression.java (hunk 9 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/grammar/RegexExpression.java]: All of these characters are regex special symbols, so they should be escaped to be treated as raw characters, and Java string literal requires backslashes to be escaped with backslashes  to represent backslash character for regex escaping.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/Parser.java (hunk 13 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/Parser.java]: Amount of initial states has nothing common with amount of active states during parsing. It is ok with default preallocation.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParserFsm.java (hunk 12 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParserFsm.java]: I don't see why do we need it.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseResult.java (hunk 18 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/ParseResult.java]: Don't catch the point.

File: plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java (hunk 2 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java]: Thanks for suggestion
